ts/import-csv/FieldMapper.svelte

```diff
@@ -3,27 +3,27 @@ Copyright: Ankitects Pty Ltd and contributors
 License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
 -->
 <script lang="ts">
+    import type { CsvMetadata_MappedNotetype } from "@tslib/anki/import_export_pb";
+    import { getFieldNames } from "@tslib/anki/notetypes_service";
     import * as tr from "@tslib/ftl";
-    import type { ImportExport } from "@tslib/proto";
 
     import Spacer from "../components/Spacer.svelte";
     import type { ColumnOption } from "./lib";
-    import { getNotetypeFields } from "./lib";
     import MapperRow from "./MapperRow.svelte";
 
     export let columnOptions: ColumnOption[];
     export let tagsColumn: number;
-    export let globalNotetype: ImportExport.CsvMetadata.MappedNotetype | null;
+    export let globalNotetype: CsvMetadata_MappedNotetype | null;
 
-    let lastNotetypeId: number | undefined = -1;
+    let lastNotetypeId: bigint | undefined = -1n;
```

Comment:
Does protobuf-es also properly support optionals so that we can move away from sentinel values?

Reply:
For lists of optionals (eg change notetype screen), Protobuf doesn't allow for repeated optionals, so the optional would still need to be placed in a wrapper message that is repeated, eg the code would need to check 'if (x[5].val !== null)' instead of 'if (x[5] !== null)'. IIRC that's the main reason we are currently using sentinels in lists.

I don't recall why a sentinel was used in the code you quoted, and undefined probably would work instead.

Reply:
If I'm not mistaken, the old library always set missing optional numbers to 0, and there was no way to distinguish it from the actual value 0.

Reply:
Ah, that could be it.